File: leet/1337_k_weakest_rows_in_matrix/1337_k_weakest_rows_in_matrix.py
# LeetCode 1337. The K Weakest Rows in a Matrix
# https://leetcode.com/problems/the-k-weakest-rows-in-a-matrix/
# O(m(logn + logk)) time, O(m) space
class Solution:
    def kWeakestRows(self, mat: List[List[int]], k: int) -> List[int]:
        from heapq import heappush, heappop
        heap = []
        def binarySearchFirstZero(row):
            lo, hi = 0, len(row)
            while lo < hi:
                mid = lo + (hi - lo) // 2
                if row[mid] == 1:
                    lo = mid + 1
                else:
                    hi = mid
            return lo
        strengths = [(binarySearchFirstZero(row), i) for i, row in enumerate(mat)]
        for strength, i in strengths:
            if (len(heap) < k or strength < -heap[0][0] or
                (strength == -heap[0][0] and i < -heap[0][1])):
                if len(heap) >= k:
                    heappop(heap)
                heappush(heap, (-strength, -i))
        return reversed([-heappop(heap)[1] for _ in range(len(heap))])

File: leet/1337_k_weakest_rows_in_matrix/1337_k_weakest_rows_in_matrix.ts
// LeetCode 1337. The K Weakest Rows in a Matrix
// https://leetcode.com/problems/the-k-weakest-rows-in-a-matrix/
// O((m + n)logn) time, O(m) space
function kWeakestRows(mat: number[][], k: number): number[] {
    const binarySearchFirstZero = (row: number[]): number => {
        let lo = 0, hi = row.length;
        while (lo < hi) {
            let mid = Math.floor(lo + (hi - lo) / 2);
            if (row[mid] == 1)
                lo = mid + 1;
            else
                hi = mid;
        }
        return lo;
    }  
    return new Array(mat.length).fill([])
        .map((a, i) => [binarySearchFirstZero(mat[i]), i])
        .sort((a, b) => (a[0] == b[0] ? a[1] - b[1] : a[0] - b[0]))
        .map(a => a[1])
        .slice(0, k);
};
